### ISSUE TITLE
Tree nodes re-render when cursor hovers over toolbar shortcut icons ( setToolbarOverlay )

### ISSUE THREAD

**Issue by shresthabijay** [cybersemics/em#527]
**Issue:**
When hovered over any one of the shortcut icons from the toolbar, an overlay describing the shortcut is rendered. But the event  is causing whole tree nodes to re-render. This side effect causes unnecessary re-renders and hinders application performance.

**Expected behaviour**
Just rendering an overlay for shortcuts in a toolbar , should not cause whole tree to re-render. I strongly believe that this event should only re-render toolbar. It doesn't cause any state changes that is consumed by `Thoughts`.

_Hovering over shortcuts icons_

![captured (3)](https://user-images.githubusercontent.com/25879100/78765050-270ec480-79a7-11ea-8709-dc15d5dc5140.gif)

**Comment by raineorshine** [comment]
I agree! Good catch.

(How do you get the render boxes to appear???)

**Comment by shresthabijay** [comment]
> (How do you get the render boxes to appear???)

Go to the react dev tool component. Click at the top right setting icon and check highlight update.
It will highlight re-rendered components. Colors vary with frequency of re-renders.
  
![higlight-update](https://user-images.githubusercontent.com/25879100/78801838-e0d45800-79dc-11ea-9d94-762adfeda41b.png)

**Comment by anmolarora1** [comment]
@raineorshine Adding to what @shresthabijay pointed out, I also noticed this issue sometime back, and many such issues can be caught using dev tools and profiler. 
Afaik, this one is happening because of `rootThoughts` is being re-calculated on every render and even if its value doesn't change, its reference does. 
![Screen Shot 2020-04-08 at 6 16 34 PM](https://user-images.githubusercontent.com/7765764/78808528-d3f83a00-79c5-11ea-997f-39e50d956d1d.png)

**Comment by shresthabijay** [comment]
@anmolarora1 good catch. @raineorshine This is one of the few places in codebase where referential inequality in `mapStateToProps` is causing unnecessary re-renders. I believe I suggested you about this when I first reviewed the em codebase.

It is better to use `mapStateToProps`  to return only state from the store that visual component depends on and do all the necessary calculations inside render function. We only use such calculations inside `mapStateToProps` if these calculation result are entirely independent to other state and affects UI. If they are dependent on any one of the state, we return those required states from the `mapStateToProps` and then do calculation on render function. This will eliminate referential inequality and keeps `mapStateToProps` simple and easy to follow. 

Lets open an issue to refactor this over required places in codebase.
